go/cmd/dolt/commands/branch.go

```diff
@@ -103,89 +106,163 @@ func (cmd BranchCmd) Exec(ctx context.Context, commandStr string, args []string,
 	branchSet := set.NewStrSet(apr.Args)
 
 	verbose := apr.Contains(cli.VerboseFlag)
 	printRemote := apr.Contains(cli.RemoteParam)
 	printAll := apr.Contains(cli.AllFlag)
 
-	branches, err := dEnv.DoltDB.GetHeadRefs(ctx)
+	var branches []branchMeta
+	if printAll || printRemote {
+		remoteBranches, err := getBranches(sqlCtx, queryEngine, true)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.BuildDError("error: failed to read remote branches from db").AddCause(err).Build(), nil)
+		}
+		branches = append(branches, remoteBranches...)
+	}
 
-	if err != nil {
-		return HandleVErrAndExitCode(errhand.BuildDError("error: failed to read refs from db").AddCause(err).Build(), nil)
+	if printAll || !printRemote {
+		localBranches, err := getBranches(sqlCtx, queryEngine, false)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.BuildDError("error: failed to read local branches from db").AddCause(err).Build(), nil)
+		}
+		branches = append(branches, localBranches...)
 	}
 
-	currentBranch, err := dEnv.RepoStateReader().CWBHeadRef()
+	currentBranch, err := getActiveBranchName(sqlCtx, queryEngine)
 	if err != nil {
-		return HandleVErrAndExitCode(errhand.BuildDError("error: failed to read refs from db").AddCause(err).Build(), nil)
+		return HandleVErrAndExitCode(errhand.BuildDError("error: failed to read current branch from db").AddCause(err).Build(), nil)
 	}
+
 	sort.Slice(branches, func(i, j int) bool {
-		return branches[i].String() < branches[j].String()
+		return branches[i].name < branches[j].name
 	})
 
 	for _, branch := range branches {
-		if branchSet.Size() > 0 && !branchSet.Contains(branch.GetPath()) {
-			continue
-		}
-
-		cs, _ := doltdb.NewCommitSpec(branch.String())
-
-		shouldPrint := false
-		switch branch.GetType() {
-		case ref.BranchRefType:
-			shouldPrint = printAll || !printRemote
-		case ref.RemoteRefType:
-			shouldPrint = printAll || printRemote
-		}
-		if !shouldPrint {
+		if branchSet.Size() > 0 && !branchSet.Contains(branch.name) {
 			continue
 		}
 
 		commitStr := ""
-		branchName := "  " + branch.GetPath()
+		branchName := "  " + branch.name
 		branchLen := len(branchName)
-		if ref.Equals(branch, currentBranch) {
-			branchName = "* " + color.GreenString(branch.GetPath())
-		} else if branch.GetType() == ref.RemoteRefType {
-			branchName = "  " + color.RedString("remotes/"+branch.GetPath())
-			branchLen += len("remotes/")
+		if branch.name == currentBranch {
+			branchName = "* " + color.GreenString(branch.name)
+		} else if branch.remote {
+			branchName = "  " + color.RedString(branch.name)
 		}
 
 		if verbose {
-			cm, err := dEnv.DoltDB.Resolve(ctx, cs, currentBranch)
-
-			if err == nil {
-				h, err := cm.HashOf()
-
-				if err != nil {
-					return HandleVErrAndExitCode(errhand.BuildDError("error: failed to hash commit").AddCause(err).Build(), nil)
-				}
-
-				commitStr = h.String()
-			}
+			commitStr = branch.hash
 		}
 
```

Comment:
Can you add a comment to future readers that this silliness is the result of the color characters in the branchName strings. This makes me very sad.

Reply:
f'ing github. I don't know why sometimes i comments on line that are different from what I want to comment on. I'm talking about this line:

`		fmtStr := fmt.Sprintf("%%s%%%ds\t%%s", 48-branchLen)`